### ISSUE TITLE
[Feature Request] Media Pack Creation for Personal Backups

### ISSUE THREAD

**Issue by Ltek** [syd711/vpin-studio#339]
idea is that studio would create a simple ZIP file (package) with all the media for a specific table. This would make backup/restore easier. People already post media packs on the forums, unfortunately they are not in a 'standard' format so this is a chance for Studio to 'set a standard'.

 could see the Media Pack including...

- all popper media
- B2S
- AltSound
- AltColor

When importing, Studio could identify what exists and present the user with checkboxes to select what they want imported (leaving unchecked items to not import)

**Comment by spaceangel1970** [comment]
I know Popper does media packs, but thats, just the popper media not the other stuff.

**Comment by syd711** [comment]
tltr: I'm not sure if and how to continue here.

As much as I like the idea of having a standarized archive format, this will again bring up the VPU guys of relasing something that can be used to distribute tables (or here table assets) on "non-official" ways. This situation makes this hobby so much more annoyingly cumbersome and although there is VBPM, it feels tedious dealing with encryption keys to backup and restore my tables.

**Comment by spaceangel1970** [comment]
Like you said this is already being done in VPBM. And this raises the issue
Matt had with his backup manager software.

**Comment by spaceangel1970** [comment]
However Popper already has the feature to create Media Packs for wheels,
videos but not anything table related.

**Comment by Ltek** [comment]
one thing I know for sure is with all things in life... If there is something frivolous, people will complain.  You just gotta ignore it most of the time. I create content... tables, FullDMD videos, wheels, etc.. I post them for free, just like everyone else does. People are allowed, by law, to use them however they want. If that means I want to ZIP mine up and back them up... or even give them to a friend - so be it.  Many authors are now packing VPX files they create with other content (wheels, pup packs, dmd videos, etc)... I did this with Tron Neon, and today with Swamp Thing. I will do it soon with another table.

**Comment by Ltek** [comment]
@syd711 looks like others are already doing it, openly... https://github.com/fordcrews/PopperTools/blob/main/backuptables.py

**Comment by leprinco** [comment]
@ltek, @syd711 please see https://github.com/Hagrou/tablePackager

**Comment by spaceangel1970** [comment]
I tried contacting the author of this program to see if he is still supporting. Ive had no reply.

**Comment by leprinco** [comment]
> I tried contacting the author of this program to see if he is still supporting. Ive had no reply.

Yes.. the github code contains additional codethat has never been released and it seems everything has been left over since 3 years. Saying that the published release works well and I have not seen any defects

**Comment by spaceangel1970** [comment]
I did try it out and it doesn't do puppacks and it extracts .pov files from
vpx but not .ini files.

**Comment by Ltek** [comment]
TablePacker code is way too old to use unmodified, 3+ years before INI files were used.

**Comment by hillelcoren** [comment]
I'm a developer, I'd be happy to help with this.

It'd be great if when you first launch the app it would compare your list of tables to a master list stored on the server. If there are table you're missing you can download (a few at a time like the online forums), more interestingly if you have tables the system doesn't have you can click a button to upload the table to make it available to everyone.

**Comment by hillelcoren** [comment]
@syd711 thanks for the details!

If the app is downloading tables from VPU's servers I could understand their concern however I'm suggesting the app be able to share tables between users. The more people who use the app the more complete the backend library will become. 

Online storage is now pretty cheap, I would think ads in the app would be able to cover the server costs. I'd be happy to build you a backend service to handle it, I'm experienced with Laravel which would be a great fit for this. 

https://docs.digitalocean.com/products/spaces/details/pricing/

**Comment by hillelcoren** [comment]
I agree that the forums are a great place for the community to share feedback/likes on tables. The desktop app could link back to the forum for people to provide feedback.

Browsers are great for forums, they're terrible for table distribution. The engineer in me finds it challenging to accept that the status quo is good enough. We could easily 10x the user experience. 

You're right, Lumen would be a much better fit for this. My limited understanding is that as long as you're non-commercial you're ok. That said, I understand not wanting to rock the boat.

**Comment by Ltek** [comment]
As a table author myself, I suggest the entire argument against this doesn't hold water. I made this FR so we could have an easy way for individuals on their own system to make a backup of each table, as an independent package (and restore when needed).  I dont want more convoluted software, I just want a single (and stellar) tool ... vPin Studio. 

Not only is a basic ZIP file a good, simple, easy, backup but there is an 'archive' use case... Often I try a table, customize the all the popper media, INI, etc... then decide I'm not in love with the table. I dont just want to deleted it because likely in the next year or two someone will do a major update to the table to improve it. If I deleted it I get to save the space but all the customization is removed too. Otherwise all the files sit there taking up space AND vPin Studio still must see/scan/filter it.  I do hide the table in Popper which helps from that side, but its not an archive method and takes up tons of drive space in the cabinet... vs it being archived onto my NAS as a ZIP file until I need it in a year or two and I want to play it... one click restore, update the VPX file to the latest, boom, done.

One huge reason people will not share... People customize darn near everything to their own liking. That is why there are many different B2S, videos, toppers, etc for nearly all tables. To think a backup would stop people going to download what -they like- is ridiculous.  Plus, the people who want curated tables/software pay for a service anyway... they will never use, or even hear of, Vpin Studio. I know this, I joined a few of those FB and Discord groups to see what they are doing. Its a subscription service and when I and a few others asked why they don't do it themselves using vPin Studio, they say "Thank you, no thank you"... they just want to pay the $100/yr or so and have it done for them. 

Dazz wants site traffic to get ad revenue, to run the site... cool with that and understand 100%.  This FR is not in any way intended to change site traffic.

**Comment by syd711** [comment]
I completely agree with you. And as you know my fingers are actual feel itchy to re-activate my packaging. The only reason I haven't done it yet, is that it is most likely that I'll get banned from VPU. So it would be a major drawback for me.

**Comment by hillelcoren** [comment]
I think the best way to do this is in collaboration with the main forums: 
- https://vpuniverse.com
- https://www.vpforums.org
- https://pinballnirvana.com

This would benefit everyone: people would have a much easier time installing new tables and we can use the app to drive more traffic to VPU and the other forums to rate and provide feedback on the tables.

We could use the [virtual-pinball-spreadsheet](https://raw.githubusercontent.com/VirtualPinballSpreadsheet/vps-db/refs/heads/main/db/vpsdb.json) as a starting point for the full dataset.

**Comment by Ltek** [comment]
@hillelcoren this thread about Backups, not downloads.

**Comment by hillelcoren** [comment]
From a code perspective it seems they're closely related, is there a reason to backup but not share tables?

**Comment by BRoadifer** [comment]
> As a table author myself, I suggest the entire argument against this doesn't hold water. I made this FR so we could have an easy way for individuals on their own system to make a backup of each table, as an independent package (and restore when needed). I dont want more convoluted software, I just want a single (and stellar) tool ... vPin Studio.
> 
> Not only is a basic ZIP file a good, simple, easy, backup but there is an 'archive' use case... Often I try a table, customize the all the popper media, INI, etc... then decide I'm not in love with the table. I dont just want to deleted it because likely in the next year or two someone will do a major update to the table to improve it. If I deleted it I get to save the space but all the customization is removed too. Otherwise all the files sit there taking up space AND vPin Studio still must see/scan/filter it. I do hide the table in Popper which helps from that side, but its not an archive method and takes up tons of drive space in the cabinet... vs it being archived onto my NAS as a ZIP file until I need it in a year or two and I want to play it... one click restore, update the VPX file to the latest, boom, done.
> 
> One huge reason people will not share... People customize darn near everything to their own liking. That is why there are many different B2S, videos, toppers, etc for nearly all tables. To think a backup would stop people going to download what -they like- is ridiculous. Plus, the people who want curated tables/software pay for a service anyway... they will never use, or even hear of, Vpin Studio. I know this, I joined a few of those FB and Discord groups to see what they are doing. Its a subscription service and when I and a few others asked why they don't do it themselves using vPin Studio, they say "Thank you, no thank you"... they just want to pay the $100/yr or so and have it done for them.
> 
> Dazz wants site traffic to get ad revenue, to run the site... cool with that and understand 100%. This FR is not in any way intended to change site traffic.

If you really think I run the site to get any revenue... You are grossly incorrect. Over the years 10 years or so, I have spent several ten's of thousands $$$$$ out of my own pocket to keep the site alive. Over the past 3 years, I added Subscriptions and ad revenue. That helps, but there is sooooooo much more going on in the background that you have no clue about. Sites that produce 70+TB bandwidth per month are not cheap to run at all...

**Comment by BRoadifer** [comment]
> I completely agree with you. And as you know my fingers are actual feel itchy to re-activate my packaging. The only reason I haven't done it yet, is that it is most likely that I'll get banned from VPU. So it would be a major drawback for me.

Most content authors that I have spoken to have been strongly against the backup packaging that you have proposed.  I've talked to several of the major ones and many are on the same page. They feel strongly that overall it would be bad for the community if these were sharable. Packs would be much easier to create, and easily enable the re-hosting/packaging for pack/drive sellers.

**Comment by Ltek** [comment]
> If you really think I run the site to get any revenue... You are grossly incorrect. Over the years 10 years or so, I have spent several ten's of thousands $$$$$ out of my own pocket to keep the site alive. Over the past 3 years, I added Subscriptions and ad revenue. That helps, but there is sooooooo much more going on in the background that you have no clue about. Sites that produce 70+TB bandwidth per month are not cheap to run at all...

You and I have never spoken but I assumed your were not doing it for personal gain. I said "Dazz wants site traffic to get **ad revenue, to run the site**... cool with that and understand 100%. This FR is not in any way intended to change site traffic."

>Most content authors that I have spoken to have been strongly against the backup packaging that you have proposed

As I stated in the prior post; most community members customize the heck out of everything (which is why we make tons of B2S, videos, wheels, etc, etc ,etc... for darn near EVERY table. 'Sharing' would be an inconsequentially  low % of people (for all the reasons I stated above - and prob more reasons I have not thought about). The people who do not customize thier own do not want to do ANY WORK AT ALL... they pay a subscription and get a completely 100% curated drive with 1000 tables pre-loaded.
Any 'reason' to disallow a tool for people to have easy backups is ridiculous. Anyone can create it this with PowerShell, VBscript or even a BAT file and a free copy of 7ZIP.
I did a [single web search ](https://www.google.com/search?q=script+to+zip+up+files+in+multiple+directories+with+the+same+name&sca_esv=3155dd6ef610aa34&sxsrf=ADLYWIIxpNHTKZqwsKWSkKF2giG5f1vMSQ%3A1733793610656&ei=SpdXZ5vQJ8iCm9cP4fC_2Qg&ved=0ahUKEwibnZvChJyKAxVIweYEHWH4L4sQ4dUDCA8&uact=5&oq=script+to+zip+up+files+in+multiple+directories+with+the+same+name&gs_lp=Egxnd3Mtd2l6LXNlcnAaAhgCIkFzY3JpcHQgdG8gemlwIHVwIGZpbGVzIGluIG11bHRpcGxlIGRpcmVjdG9yaWVzIHdpdGggdGhlIHNhbWUgbmFtZUilXVAAWMVbcAF4AZABAJgB2gGgAcU6qgEHMTIuNTMuMbgBA8gBAPgBAZgCPqAC8DfCAgQQIxgnwgIKECMYgAQYJxiKBcICChAAGIAEGEMYigXCAhAQABiABBixAxhDGIMBGIoFwgILEAAYgAQYsQMYgwHCAggQABiABBixA8ICBRAAGIAEwgINEC4YgAQYQxjUAhiKBcICERAuGIAEGLEDGNEDGIMBGMcBwgILEC4YgAQYsQMY1ALCAhEQLhiABBixAxiDARjHARivAcICCxAuGIAEGMcBGK8BwgIOEC4YgAQYsQMY0QMYxwHCAgsQLhiABBixAxiDAcICERAuGIAEGJECGNEDGMcBGIoFwgILEAAYgAQYkQIYigXCAiAQLhiABBiRAhjRAxjHARiKBRiXBRjcBBjeBBjgBNgBAcICBhAAGBYYHsICCRAAGBYYxwMYHsICCxAAGIAEGIYDGIoFwgILEAAYgAQYogQYiwPCAggQABiLAxjvBcICCBAAGBYYChgewgIFECEYoAHCAgUQIRirAsICCBAhGIsDGJ8FwgIFECEYnwXCAggQABiABBiiBJgDALoGBggBEAEYFJIHBjUuNTYuMaAH-54E&sclient=gws-wiz-serp) and found 5 fully functional code examples on the first search page.

**Comment by hillelcoren** [comment]
Although some people buy table packs and others want to endlessly customize, the vast majority of users just want something simple to use. The goal should be to make installing tables as easy to use as Steam. Requiring tables to only be shared on the forum will greatly hold back this space. 

There's only one table creator on this thread and they don't appear to have a problem sharing their tables. My understanding is that sharing tables needs to be noncommercial to avoid potential licensing issues. The issue isn't whether or not the forums make money, it's whether or not they charge money. Providing tables for free through the app seems like a better option.

The better solutions always end up winning, it's just a matter of how long it takes. The forums are in important part of the community, there's an opportunity now to work together to improve the situation for everyone.

**Comment by Ltek** [comment]
@hillelcoren please keep this thread to BACKUP... not sharing.  Also, the forums are a vital part of the ecosystem and community... we need to support Dazz and help him keep the VPU thriving (again, that is a separate, not on topic of backups)

**Comment by BRoadifer** [comment]
> Although some people buy table packs and others want to endlessly customize, the vast majority of users just want something simple to use. The goal should be to make installing tables as easy to use as Steam. Requiring tables to only be shared on the forum will greatly hold back this space.
> 
> There's only one table creator on this thread and they don't appear to have a problem sharing their tables. My understanding is that sharing tables needs to be noncommercial to avoid potential licensing issues. The issue isn't whether or not the forums make money, it's whether or not they charge money. Providing tables for free through the app seems like a better option.
> 
> The better solutions always end up winning, it's just a matter of how long it takes. The forums are in important part of the community, there's an opportunity now to work together to improve the situation for everyone.

Things are the way they are currently because of agreements that we have with the real manufactures... Newer people getting into this hobby have no clue what we have gone through in the past to get the hobby where it is today. The legit sites have had very few issues with manufactures because we follow their rules. Over 20 years we have been working and speaking with manufactures in regards to what we can and cannot do.

Their first rule and primary rule for us, 20 years ago, is that ROMs cannot be included with the table. Thus packaging a table, with the ROM, goes against the manufactures rules that we agreed with them on. 

Most of the things we do now are done to keep this hobby on a hobbyist level. Sure we can make some things easier now, but some of that would take us into a more commercial product. If that happens then we'd be headed towards expensive licensing.

With that said, I do have an idea for an app, table browser/installer that would make things easier. IE: Steam Workshop for the hobby.  VPU also has a built in API that could be used to gather and present data.

**Comment by Ltek** [comment]
@BRoadifer this FR was always about 'media'. Not VPX files, not ROMs. This is why I named it Media Pack.

> could see the Media Pack including...
> 
> all popper media
> B2S
> AltSound
> AltColor

The 2nd goal/intent is to create a structured format/methodology so Studio can recognize multiple media files and what they are for.
Example...
If you DL a media pack (ZIP) today from VPU, content creators are throwing files with random naming formats into folder structures (or not) they decide at that moment... thus, impossible for Studio to understand there could be four Help Card images that need to be added/appended in a specific order into Popper. Same for even DMD videos, or Loading videos... it doesnt know which is which, and how many of each there are.

**Comment by syd711** [comment]
I think we can agree on that 
- we want some kind of backup mechanismn (I think this statement is true, no matter if you are a table author or not)
- we do not want that these backups are used for unofficial distribution

One could argue that VPBM is out there, but there are several reasons **I am** not a 100% happy with this solution:
- It is closed source (due to it's nature)
- It is maintained by a single developer
- The Studio integration is not that seamlessly as it could be (of course this could be improved), but ...
- It is cumbersome. Personally I am simply too lazy to copy the keys between systems to set it up properly.

So here is my pragmatic suggestion to solve this:
- The Studio provides a preferences page where the VPU login data is configured.
- The Studio provides an import and export functionality of table archives which are **password protected zip files** (let's leave the internal structure of these aside for a moment).
- The Studio uses the login data as password for these zipfiles. Therefore, they are bound to a specific user and can be imported and  exported via any VPin Studio client. 
- The credentials are validated against VPU during import/export.

Benefits:
- If the unlike case happens that these archives are distributed in some way over the internet, you know exactly which account is responsible for this and can block it which would invalidate all related backups out there.
- No changes required from VPU site. This would only cause some more authorization requests.

It's an idea I came up this morning, so feel free to comment on this.
I know from technical point of view, it is not a 100% secure, but it is secure enough to protect against commercial abuse.

**Comment by spaceangel1970** [comment]
I think that would be a perfect solution and would work great for me
personally.

**Comment by syd711** [comment]
For me an archive format for media packs and for archiving tables + media should be the same. There should be **one** unified format, no matter if a VPX or ROM file is in it or not. 

Yes, for pure media packs we can leave the password protection aside. This is only a subset of an actual backup and these files/format only ensure that assets can be installed in a certain way via Studio or other tools. So I see no reason not to go ahead here as part of this FR.

If the archive contains a tables (and ROM files), the additional password strategy could be applied and **can** be part of this FR too - if @BRoadifer does not object here, which would be fine for me too.

**Comment by BRoadifer** [comment]
Sorry guys... I haven't been trying to ignore you...  I just absolutely despise github with a passion and spend as little time here as possible.  Hit me on Discord, VPU, or even Facebook. ... anywhere, but git.

Keep in mind when it comes to media... 

all popper media
B2S
AltSound
AltColor

I created the old "Media Pack" standard back when we developed HyperPin. That was almost 15 years ago....  Back then, most of the files (if not all) included in the media pack was created/re-touched by the person that released it. Initially, I created hundreds of them until other people started jumping in. 

Fast forward to today... Pinup Popper media server handles the bulk of the media that we considered "front end media" in the past. B2S, AltSound, and AltColor are NOT "front end media".  These are NOT parts of the Popper media because the majority of these files are created by many different individuals. These are files take more time to create than any piece of "front end media" ever could. AltColor alone is a HUGE project and takes forever to create. Tens to hundreds of hours could be involved in that creation. I do not think it would be fair to the authors of those file types to be included in a "media pack".

Now, with saying all of that... Invision Community which VPUniverse uses has a built in API.  https://invisioncommunity.com/developers/rest-api/index/  I am not familiar with everything it can do, but perhaps it could be of some use. If that could be of any use, we can figure something out.  

Honestly, I don't plan on keeping VPU in this format forever and looking into going a different route that would be better for everyone eventually.  I'm not a developer, and it's very difficult for me to trust just anyone... I've spent 20 years getting the communities trust and getting VPU to where it currently is.

**Comment by Ltek** [comment]
@BRoadifer you do a great job with VPU, the best forum I've seen for vPins. I'm an engineer so my brain always things good, can be better ;-) so very curious what your new 'route' might look like. I prefer forums 1000x over the 'gone in a moment' options like discord/facebook/etc... those are very bad for long multi-user conversations, asks for help, historical documents/learning/searching, etc.

**Comment by syd711** [comment]
Fixed with 4.2
